rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -993,19 +1008,102 @@ impl SystemState {
             CanisterStatus::Running {
                 call_context_manager,
-            } => Some(call_context_manager),
+            } => Ok(call_context_manager),
             CanisterStatus::Stopping {
                 call_context_manager,
                 ..
-            } => Some(call_context_manager),
-            CanisterStatus::Stopped => None,
+            } => Ok(call_context_manager),
+            CanisterStatus::Stopped => Err(StateError::CanisterStopped(self.canister_id)),
         }
     }
 
+    /// Creates a new call context and returns its ID. Returns an error if the
+    /// canister is `Stopped`.
+    pub fn new_call_context(
+        &mut self,
+        call_origin: CallOrigin,
+        cycles: Cycles,
+        time: Time,
+        metadata: RequestMetadata,
+    ) -> Result<CallContextId, StateError> {
+        Ok(self
+            .call_context_manager_mut()?
+            .new_call_context(call_origin, cycles, time, metadata))
+    }
+
+    /// Withdraws cycles from the call context with the given ID.
+    ///
+    /// Returns a reference to the `CallContext` if successful. Returns an error
+    /// message if the canister is `Stopped`; the call context does not exist; or
+    /// if the call context does not have enough cycles.
+    pub fn withdraw_cycles(
+        &mut self,
+        call_context_id: CallContextId,
+        cycles: Cycles,
+    ) -> Result<&CallContext, &str> {
+        self.call_context_manager_mut()
+            .map_err(|_| "Canister is stopped")?
+            .withdraw_cycles(call_context_id, cycles)
+    }
+
+    /// Accepts a canister result for the given `CallContext` and produces an action
+    /// that should be taken by the caller; and the call context, if completed.
+    pub fn on_canister_result(
+        &mut self,
+        call_context_id: CallContextId,
+        result: Result<Option<WasmResult>, HypervisorError>,
+        instructions_used: NumInstructions,
+    ) -> Result<(CallContextAction, Option<CallContext>), StateError> {
+        Ok(self.call_context_manager_mut()?.on_canister_result(
+            call_context_id,
+            result,
+            instructions_used,
+        ))
+    }
+
+    /// Marks all call contexts as deleted and produces reject responses for the
+    /// not yet responded ones. This is called as part of uninstalling a canister.
+    ///
+    /// Callbacks will be unregistered when responses are received.
+    pub fn delete_all_call_contexts<R>(
+        &mut self,
+        reject: impl Fn(&CallContext) -> Option<R>,
+    ) -> Vec<R> {
+        if let Ok(call_context_manager) = self.call_context_manager_mut() {
+            call_context_manager.delete_all_call_contexts(reject)
+        } else {
+            Vec::new()
+        }
+    }
+
+    /// Registers a callback and returns its ID. Returns an error if the canister is
+    /// `Stopped`.
+    //
+    // TODO: Check whether this could be done implicitly, when pushing an outbound
+    // request.
```

Comment:
I was wondering exactly this. Are there possible use-cases for registering a callback without a message? Seems to me like this should go hand in hand with pushing an outbound request.

Reply:
And then on top of that how about also unregistering when pushing an inbound response? Perhaps it would make sense to have a separate set of 'lingering callbacks' or something just for DTS? Something that wouldn't cause problems with the new messaging model but allows for this requirement.

Reply:
I'm not sure that it's practical (without significant refactoring of Execution code) to bundle pushing or popping this or that message with registering or unregistering a callback / call context. Apart from callbacks and DTS execution of responses, it's quite possible that Execution forgoes creating a call context when a request / ingress is executed in one go.

But it can definitely be looked into. And removing the callback and temporarily keeping it on the side upon popping of a response is indeed a simpler solution than implementing a state machine (assuming that this is e.g. one of two exceptions to the rule of messages and callbacks/call contexts going hand in hand; otherwise it can quickly get out of hand).